# **Second section**


*   Hacer un análisis de calidad de datos sobre *dataset.csv*
*   Generar un reporte de los resultados.

### Note:  encontrar todas las anomalías de calidad de datos del dataset. No deberá corregir las anomalías que encuentre, solo encontrarlas y justificarlas en el reporte.
Entregable: Reporte de calidad de datos del dataset, adicionalmente puede incluir el código que muestre cómo encontró las anomalías de calidad de datos.

In [18]:
## Importing libraries

import pandas as pd
import json
import matplotlib.pyplot as plt
import re

In [2]:
## Mounting Google Drive in the Colab environment:

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
# Path to the CSV file
file_path = '/content/drive/My Drive/Python/R5/dataset.csv'

# Read the CSV file into a DataFrame named 'dataset'
df = pd.read_csv(file_path)

df.head(3)

disc_number  duration_ms explicit  track_number  track_popularity  \
0            1       212600    False             1                77   
1            1       231833    False             2                78   
2            1       231000    False             3                79   

                 track_id                              track_name  \
0  4WUepByoeqcedHoYhSNHRt  Welcome To New York (Taylor's Version)   
1  0108kcWLnn2HlH2kedi1gn          Blank Space (Taylor's Version)   
2  3Vpk1hfMAQme8VJ0SNRSkd                Style (Taylor's Version)   

   audio_features.danceability  audio_features.energy  audio_features.key  \
0                        0.757                  0.610                 7.0   
1                        0.733                  0.733                 0.0   
2                        0.511                  0.822                11.0   

   ...  audio_features.tempo       audio_features.id  \
0  ...               116.998  4WUepByoeqcedHoYhSNHRt   
1  ...                96.057  0108kcWLnn2HlH2kedi1gn   
2  ...                94.868  3Vpk1hfMAQme8VJ0SNRSkd   

   audio_features.time_signature           artist_id   artist_name  \
0                            4.0  06HL4z0CvFAxyc27GX  Taylor Swift   
1                            4.0  06HL4z0CvFAxyc27GX  Taylor Swift   
2                            4.0  06HL4z0CvFAxyc27GX  Taylor Swift   

   artist_popularity                album_id  \
0                120  1o59UpKw81iHR0HPiSkJR0   
1                120  1o59UpKw81iHR0HPiSkJR0   
2                120  1o59UpKw81iHR0HPiSkJR0   

                         album_name album_release_date  album_total_tracks  
0  1989 (Taylor's Version) [Deluxe]         2023-10-27                  22  
1  1989 (Taylor's Version) [Deluxe]         2023-10-27                  22  
2  1989 (Taylor's Version) [Deluxe]         2023-10-27                  22  

[3 rows x 27 columns]

# 1. Completeness: Checking for any null or missing values

#### Analysis

In [4]:
## Calculating the percentage of missing values for each column
total_rows = len(df)
percentage_missing = (df.isnull().sum() / total_rows) * 100

## Creating a new DataFrame with both the count and percentage of missing values
completeness_report = pd.DataFrame({
    'Missing Values': df.isnull().sum(),
    'Percentage of Missing Values': percentage_missing
})

completeness_report

Missing Values  Percentage of Missing Values
disc_number                                   0                      0.000000
duration_ms                                   0                      0.000000
explicit                                      0                      0.000000
track_number                                  0                      0.000000
track_popularity                              0                      0.000000
track_id                                      8                      1.484230
track_name                                    7                      1.298701
audio_features.danceability                   2                      0.371058
audio_features.energy                         2                      0.371058
audio_features.key                            1                      0.185529
audio_features.loudness                       2                      0.371058
audio_features.mode                           0                      0.000000
audio_features.speechiness                    1                      0.185529
audio_features.acousticness                   1                      0.185529
audio_features.instrumentalness               0                      0.000000
audio_features.liveness                       1                      0.185529
audio_features.valence                        0                      0.000000
audio_features.tempo                          1                      0.185529
audio_features.id                             0                      0.000000
audio_features.time_signature                 1                      0.185529
artist_id                                     0                      0.000000
artist_name                                   0                      0.000000
artist_popularity                             0                      0.000000
album_id                                      0                      0.000000
album_name                                   62                     11.502783
album_release_date                            0                      0.000000
album_total_tracks                            0                      0.000000

In [5]:
## Generate the report
print("The dataset has missing values in several columns:\n")
for index, row in completeness_report[completeness_report['Missing Values'] > 0].iterrows():
  print("\t"f"{index}: {int(row['Missing Values'])} missing values ({row['Percentage of Missing Values']:.2f}%)")

The dataset has missing values in several columns:

	track_id: 8 missing values (1.48%)
	track_name: 7 missing values (1.30%)
	audio_features.danceability: 2 missing values (0.37%)
	audio_features.energy: 2 missing values (0.37%)
	audio_features.key: 1 missing values (0.19%)
	audio_features.loudness: 2 missing values (0.37%)
	audio_features.speechiness: 1 missing values (0.19%)
	audio_features.acousticness: 1 missing values (0.19%)
	audio_features.liveness: 1 missing values (0.19%)
	audio_features.tempo: 1 missing values (0.19%)
	audio_features.time_signature: 1 missing values (0.19%)
	album_name: 62 missing values (11.50%)


#### Report

The dataset has missing values in several columns:

	track_id: 8 missing values (1.48%)
	track_name: 7 missing values (1.30%)
	audio_features.danceability: 2 missing values (0.37%)
	audio_features.energy: 2 missing values (0.37%)
	audio_features.key: 1 missing values (0.19%)
	audio_features.loudness: 2 missing values (0.37%)
	audio_features.speechiness: 1 missing values (0.19%)
	audio_features.acousticness: 1 missing values (0.19%)
	audio_features.liveness: 1 missing values (0.19%)
	audio_features.tempo: 1 missing values (0.19%)
	audio_features.time_signature: 1 missing values (0.19%)
	album_name: 62 missing values (11.50%)

# 2. Consistency: Examine the dataset for any inconsistencies (**Logical** inconsistencies in this case.)

#### Analysis

In [16]:
# Calculate the number of unique values in each column
unique_value_counts = df.nunique()

print(unique_value_counts)

disc_number                          2
duration_ms                        364
explicit                             4
track_number                        46
track_popularity                    73
track_id                           512
track_name                         331
audio_features.danceability        267
audio_features.energy              348
audio_features.key                  12
audio_features.loudness            448
audio_features.mode                  2
audio_features.speechiness         292
audio_features.acousticness        401
audio_features.instrumentalness    240
audio_features.liveness            271
audio_features.valence             326
audio_features.tempo               450
audio_features.id                  519
audio_features.time_signature        3
artist_id                            1
artist_name                          1
artist_popularity                    1
album_id                            26
album_name                          24
album_release_date       

In [17]:
# Apply value_counts() to each column
for column in df.columns:
    counts = df[column].value_counts().sort_index()
    # Mostrar todos los valores y recuentos
    # with pd.option_context('display.max_rows', None):
    print(f"Value counts for {column}:\n{counts}\n\n\n")

Value counts for disc_number:
1    522
2     17
Name: disc_number, dtype: int64



Value counts for duration_ms:
-223093    1
-107133    1
 10        1
 1000      1
 3000      1
          ..
 403887    1
 403933    1
 404680    1
 405906    1
 613026    1
Name: duration_ms, Length: 364, dtype: int64



Value counts for explicit:
False    480
No         4
Si         1
True      54
Name: explicit, dtype: int64



Value counts for track_number:
1     28
2     28
3     28
4     28
5     28
6     28
7     28
8     29
9     27
10    27
11    27
12    27
13    27
14    25
15    24
16    21
17    18
18    14
19    12
20    10
21    10
22     8
23     4
24     3
25     3
26     3
27     2
28     2
29     2
30     2
31     1
32     1
33     1
34     1
35     1
36     1
37     1
38     1
39     1
40     1
41     1
42     1
43     1
44     1
45     1
46     1
Name: track_number, dtype: int64



Value counts for track_popularity:
-92     1
-85     1
-75     1
-71     1
-70     1
       ..
 91     3

In [45]:
## Function to check if a track name contains unusual characters (assuming r'[^a-zA-Z0-9\s\'\’\"\‘\!?\-().,\&]' are usual characters)
def contains_unusual_characters(name):
    # Define the regex for allowed characters: letters, spaces, hyphens, and apostrophes
    if re.search(r'[^a-zA-Z0-9\s\'\’\"\‘\!?\-().,\&\/]', str(name)):
        return True
    return False

# Apply the function to the 'track_name' column
df['unusual_characters'] = df['track_name'].apply(contains_unusual_characters)

# Filter and display rows with unusual characters
unusual_names_df = df[df['unusual_characters']]

# Display results
unusual_names_df

Empty DataFrame
Columns: [disc_number, duration_ms, explicit, track_number, track_popularity, track_id, track_name, audio_features.danceability, audio_features.energy, audio_features.key, audio_features.loudness, audio_features.mode, audio_features.speechiness, audio_features.acousticness, audio_features.instrumentalness, audio_features.liveness, audio_features.valence, audio_features.tempo, audio_features.id, audio_features.time_signature, artist_id, artist_name, artist_popularity, album_id, album_name, album_release_date, album_total_tracks, unusual_characters]
Index: []

[0 rows x 28 columns]

In [49]:
## Create a new column that combines 'album_id' and 'album_total_tracks'
df['album_id_tracks_combo'] = df['album_id'].astype(str) + "_" + df['album_total_tracks'].astype(str)

# Perform a value_counts on the new column
combo_counts = df['album_id_tracks_combo'].value_counts().sort_index()

# Display the results
combo_counts

08CWGiv27MVQhYpuTtvx83_16          16
0PZ7lAru5FDFHuirTkWe9Z_34          34
151w1FgRZfnKZA9FEcg9Z3_13          13
1KVKqWeRuXsJDLTW0VuD29_22          22
1MPAXuTVL2Ej5x0JHiSPq8_46          46
1NAmidJlEaVgA3MpcPFYGq_18          36
1fnJ7k0bllNfL1kVdNVW1A_24          24
1o59UpKw81iHR0HPiSkJR0_22          22
1pzvBxYgT6OVwJLtHkrdQK_17          17
1yGbNOtRIgdIiGHOEBaZWf_19          19
1ycoesYxIFymXWebfmz828_8            8
2Xoteh7uEpea4TohMxjtaq_10          15
2fenSS68JI1h4Fo296JfGr_16          16
2gP2LMVcIFgVczSJqn340t_19          19
3lS1y25WAhcqJDATJK70Mq_20          20
4hDok0OAJd57SGIT8xuWJH_26          26
5AEDGbliTTfjOB8TSm1sxt_22          22
5eyZZoQEFQWRHkV2xgAeBw_Thirteen    15
5fy0X0JmZRZnVa2UEicIOl_13          13
64LU4c1nfjz1t4VnGhagcg_21          21
6AORtDjduMM3bupSWzbTSG_17          17
6Ar2o9KCqcyYF9J0aQP3au_14          14
6DEjYFkNZh67HP7R9PSZvv_15          16
6S6JQWzUrJVcJLK4fi74Fw_22          22
6fyR4wBPwLHKcRtxgd4sGh_16          16
6kZ42qRrzov54LcAk4onW9_34          30
Name: album_

#### Report

* The **duration_ms** column, which presumably represents the duration of tracks in milliseconds, exhibits a range of values with some notable observations:






> 1.   There are negative values (e.g., -223093, -107133), which are likely data entry errors as duration cannot be negative.


> 2.   Extremely low values (e.g., 10 ms, 1000 ms) are present. These are unusually short for typical track durations and might represent incorrect or placeholder data.
> 3. The range of durations extends up to 613026 milliseconds (or approximately 10.2 minutes), indicating the presence of longer tracks.

> > Recommendations:


> > 1.   Data Cleaning: Negative and extremely low values should be investigated and corrected if they are errors.
> > 2.   Data Validation: It is important to validate the duration_ms data against expected norms or external benchmarks for track lengths to ensure accuracy.
> > 3. Data Analysis Considerations: For any analysis involving track duration, consider handling outliers to avoid skewed results.





*   The **explicit** column has values like 'False', 'True', 'Si', 'No'. The presence of both English and Spanish values ('Si', 'No') for a boolean field indicates inconsistency.

* The set of possible values for the **track_popularity** field is presumably {0, 1, 2, 3, ..., 100}. If this is true, all different values imply data inconsistency.

* The **audio_features.acousticness** column, which presumably measures the acousticness of tracks, shows a range of values with a couple of noteworthy points:

> 1. There are negative values present (e.g., -0.003540, -0.000537). Typically, acousticness measures range from 0 to 1, where 0 represents less acoustic and 1 represents highly acoustic. Negative values could indicate data entry errors or issues with data extraction.


> 2. Values exceeding the typical range of 0 to 1 are observed, such as 1.5, 2.0, and 5.0. These are outside the standard scale for acousticness and might represent incorrect data or a different scale.

* **album_release_date**: Verify the future release dates to ensure they are intentional and not errors. This is particularly important for dates like '2027-05-26'. Also, check carefully if old dates like '1989-10-24' are correct, because it looks like an outlier.

* **album_total_tracks**: An anomaly is observed with the value 'Thirteen', which is a textual representation of the number 13. This inconsistency in data format (mix of numerical and textual data) indicates a potential data quality issue. On the other hand, the following albums appear to have an inconsistency in the album_total_tracks field, as their track count does not match the value of their field: 1NAmidJlEaVgA3MpcPFYGq, 2Xoteh7uEpea4TohMxjtaq, 5eyZZoQEFQWRHkV2xgAeBw, 6DEjYFkNZh67HP7R9PSZvv, 6kZ42qRrzov54LcAk4onW9.

# 3. Conformity: Validate that all data is in the correct format and adheres to specific standards.

#### Analysis

In [47]:
# Checking for conformity in date fields
date_conformity_check = pd.to_datetime(df['album_release_date'], errors='coerce').isna().sum()

date_conformity_check

0

#### Report

The dataset generally conforms to expected formats.

# 4. Accuracy: This dimension is challenging to assess without an external source of truth but will look for indicators of inaccuracy.

Accuracy cannot be fully assessed without external validation. However, no internal inconsistencies suggesting inaccuracy were found in the provided data.

# 5. Integrity: Assess the integrity of relationships within the dataset.


#### Analysis

In [50]:
# Check if values in 'track_id' and 'audio_features.id' match
df['integrity_check'] = df['track_id'] == df['audio_features.id']

# Count the number of matches and mismatches
integrity_result = df['integrity_check'].value_counts()

# Display the results
integrity_result

True     531
False      8
Name: integrity_check, dtype: int64

#### Report

The integrity check revealed that **track_id** and **audio_features.id** do not always match. In fact, the 8 times that do not match are the same 8 times that the track_id field has *null*. This indicates a potential issue with data integrity in the dataset.

# 6. Timeliness: Evaluate the relevance and currency of the data, particularly focusing on date fields.


#### Report

The dataset contains an album release date in the future ('2027-05-26'). This could be valid if the data is intended to include future releases, but it might also be an error if the dataset should only contain past or present data.

## Final disclaimer

I received assistance from ChatGPT, an artificial intelligence language model developed by OpenAI.